src/ahp/prover.rs

```diff
@@ -456,17 +465,32 @@ impl<F: PrimeField> AHPForR1CS<F> {
         } = *ver_message;
 
         let summed_z_m_poly_time = start_timer!(|| "Compute z_m poly");
-        let (z_a_poly, z_b_poly) = state.mz_polys.as_ref().unwrap();
-        let z_c_poly = z_a_poly.polynomial() * z_b_poly.polynomial();
+
+        let (r_a, r_b) = state.mz_rands.clone().unwrap();
+        let (randomized_z_a_poly, randomized_z_b_poly) = state.mz_rand_polys.as_ref().unwrap();
+
+        // Recover the non-randomized interpolation of z_a and z_b over H
+        let v_H: DensePolynomial<F> = domain_h.vanishing_polynomial().into();
+        let z_a_poly = randomized_z_a_poly.polynomial() - &(&v_H * r_a);
```

Comment:
Oh hm I meant that we should instead mutate `randomized_z_a_poly` in place temporarily. Also, we shouldn't convert `v_H` into a dense polynomial, as that wastes memory. Instead, we should add a way to subtract a sparse polynomial from a dense polynomial (if that doesn't already exist)

Reply:
(and also a way to multiply a sparse polynomial by a scalar, mirroring the one for `DensePolynomial`

Reply:
Gotcha, so the problem with that is `randomized_z_a_poly` and `randomized_z_b_poly` are inside an `Rc` that is borrowed by `ProverFirstOracles` [here](https://github.com/arkworks-rs/marlin/blob/6e10d143d0de7c05680da8b55ea8d39cd66da063/src/ahp/prover.rs#L389) until the end of the protocol. So we could mutate in place, but I believe this would require unsafe to do (which I'm fine doing because we know the other reference isn't used until later) - or am I missing something?

I'll make the other changes mentioned for now though!

Reply:
Hmm that’s annoying...